src/backend/distributed/citus--8.3-1--8.4-1.sql

```diff
@@ -1,6 +1,105 @@
 /* citus--8.3-1--8.4-1 */
 
 /* bump version to 8.4-1 */
+CREATE SCHEMA IF NOT EXISTS citus_internal;
+
+-- move citus internal functions to citus_internal to make space in the citus schema for
+-- our public interface
+ALTER FUNCTION citus.find_groupid_for_node SET SCHEMA citus_internal;
+ALTER FUNCTION citus.pg_dist_node_trigger_func SET SCHEMA citus_internal;
+ALTER FUNCTION citus.pg_dist_shard_placement_trigger_func SET SCHEMA citus_internal;
+ALTER FUNCTION citus.refresh_isolation_tester_prepared_statement SET SCHEMA citus_internal;
+ALTER FUNCTION citus.replace_isolation_tester_func SET SCHEMA citus_internal;
+ALTER FUNCTION citus.restore_isolation_tester_func SET SCHEMA citus_internal;
+
+CREATE OR REPLACE FUNCTION citus_internal.pg_dist_shard_placement_trigger_func()
+RETURNS TRIGGER AS $$
+  BEGIN
+    IF (TG_OP = 'DELETE') THEN
+      DELETE FROM pg_dist_placement WHERE placementid = OLD.placementid;
+      RETURN OLD;
+    ELSIF (TG_OP = 'UPDATE') THEN
+      UPDATE pg_dist_placement
+        SET shardid = NEW.shardid, shardstate = NEW.shardstate,
+            shardlength = NEW.shardlength, placementid = NEW.placementid,
+            groupid = citus_internal.find_groupid_for_node(NEW.nodename, NEW.nodeport)
+        WHERE placementid = OLD.placementid;
+      RETURN NEW;
+    ELSIF (TG_OP = 'INSERT') THEN
+      INSERT INTO pg_dist_placement
+        (placementid, shardid, shardstate, shardlength, groupid)
+      VALUES (NEW.placementid, NEW.shardid, NEW.shardstate, NEW.shardlength,
+        citus_internal.find_groupid_for_node(NEW.nodename, NEW.nodeport));
+      RETURN NEW;
+    END IF;
+  END;
+$$ LANGUAGE plpgsql;
+
+CREATE TABLE citus.pg_dist_object (
+    classid oid NOT NULL,
+    objid oid NOT NULL,
+    objsubid integer NOT NULL,
+
+    -- fields used for upgrades
+    type text DEFAULT NULL,
+    object_names text[] DEFAULT NULL,
+    object_args text[] DEFAULT NULL
+);
```

Comment:
this table, seems like it should have `PRIMARY KEY (classid, objid, objsubid)`

Reply:
I think `PRIMAREY KEY` might cause issues with upgrades.

Say pg 10 has object `A` and object `B`
Now we upgrade to PG 11 and the new address of `B` is the old address of `A`

Would this work with the `UPDATE` statement we have in `citus_finish_pg_upgrade`?

Reply:
That seems like a job for [deferred constraints](https://begriffs.com/posts/2017-08-27-deferrable-sql-constraints.html).

e.g. this works:
```sql
CREATE TABLE test (x int, y int);
INSERT INTO test VALUES (1, 0);
INSERT INTO test VALUES (-1, 1);

ALTER TABLE test ADD CONSTRAINT test_pk PRIMARY KEY (x) DEFERRABLE ;

CREATE OR REPLACE FUNCTION public.invert()
RETURNS void
LANGUAGE plpgsql
AS $function$
DECLARE
BEGIN
    SET CONSTRAINTS test_pk DEFERRED;
    UPDATE test SET x = -x;
END;
$function$;

SELECT invert();
```

Reply:
Looked promising, but postgres doesn't allow deferred on catalogs:
https://github.com/postgres/postgres/blob/292ae8af79b4f1b09a327d39e80ef70943a28194/src/backend/catalog/indexing.c#L125-L133

Assert failed when we call `CatalogTupleInsert` in `MarkObjectDistributed`

Reply:
That seems like further evidence that we should be using SPI to do an upsert / ON CONFLICT DO NOTHING there to prevent duplicates.

Reply:
actually, not quite `ON CONFLICT does not support deferrable unique constraints/exclusion constraints as arbiters`. Maybe just rewrite the update to do a DELETE+INSERT in a CTE then.

Reply:
Something like https://github.com/citusdata/citus/commit/5fc4e5f644d1ba8bca5fdc263543ef8e7b1dad86

Reply:
After that commit two create_distributed_table calls can still distribute the same schema concurrently, but it will not result in duplicate records, just some redundant CREATE SCHEMA IF NOT EXISTS call. The latter could be prevented by checking the result of the upsert. If no row was inserted then no distribution is required.

Reply:
Do we need to use SPI if we don't need `DEFERRABLE` on the constraint?
Guess we can just insert the tuple with `InsertCatalogTuple` this might give a conflict now with 2 concurrent `create_distributed_table` calls. Guess we can solve that via an appropriate lock before collecting the dependencies.

Reply:
The SPI part is to do an upsert without jumping through hoops, since concurrent inserts would error out. It did not solve the deferred constraint problem. That's solved by doing a delete+insert on upgrade.

Reply:
This is now implemented via `SPI`. This helps for concurrent `CREATE SCHEMA ... IF NOT EXISTS` that would cause uniqueness violation errors when called concurrently.

`CREATE SCHEMA` would check if it existed only once and insert later without double checking if it was already created concurrently. We guard this now with an upsert into `pg_dist_object` from the coordinator.